src/common/gpio_motor.rs

```diff
@@ -0,0 +1,411 @@
+//!         subtype: "motor".to_string(),
+//!         name: "left-motor"
+//!     },
+//!     ResourceType::Motor(Arc::new(Mutex::new(motor))),
+//! );
+//!
+//! let robot_with_motor = LocalRobot(res);
+//!
+//! ```
+//!
+
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use anyhow::Context;
+
+use super::board::{Board, BoardType};
+use super::config::ConfigType;
+use super::encoder::{
+    Encoder, EncoderPositionType, EncoderType, COMPONENT_NAME as EncoderCompName,
+};
+use super::math_utils::go_for_math;
+use super::motor::{
+    Motor, MotorPinType, MotorPinsConfig, MotorType, COMPONENT_NAME as MotorCompName,
+};
+use super::registry::{get_board_from_dependencies, ComponentRegistry, Dependency, ResourceKey};
+use super::robot::Resource;
+use super::status::Status;
+use super::stop::Stoppable;
+use crate::google;
+
+pub(crate) fn register_models(registry: &mut ComponentRegistry) {
+    if registry
+        .register_motor("gpio", &gpio_motor_from_config)
+        .is_err()
+    {
+        log::error!("gpio model is already registered")
+    }
+    if registry
+        .register_dependency_getter(
+            MotorCompName,
+            "gpio",
+            &PwmABMotor::<BoardType>::dependencies_from_config,
+        )
+        .is_err()
+    {
+        log::error!("failed to register dependency getter for gpio model")
+    }
+}
+
+// Generates a motor or an encoded motor depending on whether an encoder has been added as
+// a dependency from the config.
+pub fn gpio_motor_from_config(cfg: ConfigType, deps: Vec<Dependency>) -> anyhow::Result<MotorType> {
+    let mut enc: Option<EncoderType> = None;
+    for Dependency(_, dep) in &deps {
+        match dep {
+            Resource::Encoder(found_enc) => {
+                enc = Some(found_enc.clone());
+                break;
+            }
+            _ => {
+                continue;
+            }
+        };
+    }
+    let board = get_board_from_dependencies(deps)
+        .context("gpio motor did not receive board from dependencies")?;
+    let motor_type = if let Ok(pin_cfg) = cfg.get_attribute::<MotorPinsConfig>("pins") {
+        pin_cfg.detect_motor_type()?
+    } else {
+        return Err(anyhow::anyhow!("pin parameters for motor not found"));
+    };
+    let motor = match motor_type {
+        MotorPinType::PwmAB => PwmABMotor::<BoardType>::from_config(cfg, board.clone())?.clone(),
+        MotorPinType::PwmDirection => {
+            PwmDirectionMotor::<BoardType>::from_config(cfg, board.clone())?.clone()
+        }
+    };
+    if let Some(enc) = enc {
+        let enc_motor = EncodedMotor::new(motor, enc.clone());
+        return Ok(Arc::new(Mutex::new(enc_motor)));
+    }
+    Ok(motor)
+}
+
+// Motors generally don't care about the PWM frequency, so long as
+// it is in the order of magnitude 0f 10s of kHZ. For simplicity, we
+// just select 10kHz.
+const MOTOR_PWM_FREQUENCY: u64 = 10000;
+
+pub struct EncodedMotor<M, Enc> {
```

Comment:
discussed offline, going to do pub(crate) for everything except EncodedMotor (so that people can better test wrapping their custom encoder/motor implementations if needed)